README.md
<p align="center">
  <img src="https://images.atomist.com/sdm/SDM-Logo-Dark.png">
</p>

# @atomist/k8s-sdm

[![npm version](https://img.shields.io/npm/v/@atomist/k8s-sdm.svg)](https://www.npmjs.com/package/@atomist/k8s-sdm)
[![Docker Pulls](https://img.shields.io/docker/pulls/atomist/k8s-sdm.svg)](https://hub.docker.com/r/atomist/k8s-sdm/)

This repository contains a software delivery machine (SDM) for
deploying applications to Kubernetes using the [Atomist][atomist] API.
Currently, deploying Docker images as deployments with optional
services and ingress rules is supported.

Software delivery machines enable you to control your delivery process
in code.  Think of it as an API for your software delivery.  See the
[Atomist documentation][atomist-doc] for more information on the
concept of a software delivery machine and how to create and develop
an SDM.

[atomist-doc]: https://docs.atomist.com/ (Atomist Documentation)

## Getting started

See the [Developer Quick Start][atomist-quick] to jump straight to
creating an SDM.

[atomist-quick]: https://docs.atomist.com/quick-start/ (Atomist - Developer Quick Start)

## Contributing

Contributions to this project from community members are encouraged
and appreciated. Please review the [Contributing
Guidelines](CONTRIBUTING.md) for more information. Also see the
[Development](#development) section in this document.

## Code of conduct

This project is governed by the [Code of
Conduct](CODE_OF_CONDUCT.md). You are expected to act in accordance
with this code by participating. Please report any unacceptable
behavior to code-of-conduct@atomist.com.

## Documentation

Please see [docs.atomist.com][atomist-doc] for
[developer][atomist-doc-sdm] documentation.

[atomist-doc-sdm]: https://docs.atomist.com/developer/sdm/ (Atomist Documentation - SDM Developer)

## Connect

Follow [@atomist][atomist-twitter] and [The Composition][atomist-blog]
blog related to SDM.

[atomist-twitter]: https://twitter.com/atomist (Atomist on Twitter)
[atomist-blog]: https://the-composition.com/ (The Composition - The Official Atomist Blog)

## Support

General support questions should be discussed in the `#help`
channel in the [Atomist community Slack workspace][slack].

If you find a problem, please create an [issue][].

[issue]: https://github.com/atomist/k8s-sdm/issues

## Development

You will need to install [Node.js][node] to build and test this
project.

[node]: https://nodejs.org/ (Node.js)

### Build and test

Install dependencies.

```
$ npm install
```

Use the `build` package script to compile, test, lint, and build the
documentation.

```
$ npm run build
```

### Release

Releases are handled via the [Atomist SDM][atomist-sdm].  Just press
the 'Approve' button in the Atomist dashboard or Slack.

[atomist-sdm]: https://github.com/atomist/atomist-sdm (Atomist Software Delivery Machine)

---

Created by [Atomist][atomist].
Need Help?  [Join our Slack workspace][slack].

[atomist]: https://atomist.com/ (Atomist - How Teams Deliver Software)
[slack]: https://join.atomist.com/ (Atomist Community Slack)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased](https://github.com/atomist/k8s-sdm/compare/2.0.1...HEAD)

## [2.0.1](https://github.com/atomist/k8s-sdm/compare/2.0.0...2.0.1) - 2021-02-12

### Changed

-   Update dependencies.

## [2.0.0](https://github.com/atomist/k8s-sdm/compare/1.7.0...2.0.0) - 2020-09-04

### Added

-   Update sdm to support GitHub app provider. [451399e](https://github.com/atomist/k8s-sdm/commit/451399e2733bbcbf80355e62145f2ed3158d97e4)

### Changed

-   Update Atomist deps, remove deprecated deps. [c0b3254](https://github.com/atomist/k8s-sdm/commit/c0b3254b308e35d8bfc2a475ea4aaae9815f855d)

### Removed

-   Remove k8s cluster registration. [6264fd7](https://github.com/atomist/k8s-sdm/commit/6264fd79862bf69d3ad5f8341e168791af476fe0)

### Fixed

-   Update sdm for sync repo creds fix. [dd4d920](https://github.com/atomist/k8s-sdm/commit/dd4d920e65f4f1d11151521838f273de578f4ac6)
-   Update sdm to move sync repo creds to worker. [d06a3e0](https://github.com/atomist/k8s-sdm/commit/d06a3e04886b96d798c2a65d47262ecbb7d86743)

## [1.7.0](https://github.com/atomist/k8s-sdm/compare/1.6.1...1.7.0) - 2020-03-04

### Added

-   Create fulfiller sdm. [b57f514](https://github.com/atomist/k8s-sdm/commit/b57f514964ccc8f15242f66c1b408de07ffcacdb)
-   Add spec for sync under gvisor sandbox. [0243457](https://github.com/atomist/k8s-sdm/commit/0243457a0a836c4d50e4cacc9a691c0787ff6312)
-   Add runtimeClassName to gvisor spec. [0cff79a](https://github.com/atomist/k8s-sdm/commit/0cff79a6eb07a9cd1434431ef128c3577cf9de7a)
-   Update Atomist dependencies and changelog. [2083efa](https://github.com/atomist/k8s-sdm/commit/2083efa6d8f41cd124caa21885823fa6a593df72)

### Changed

-   Give access to namespace when namespace-scoped. [d77b829](https://github.com/atomist/k8s-sdm/commit/d77b829ff90446db5da8f52723ade2a4b860ce22)
-   Set default workers to 1. [4f79a26](https://github.com/atomist/k8s-sdm/commit/4f79a26c555ddc8b1de7831e6b17b6b5bcdd25d4)
-   Expose fulfilled goals in atomist.yaml. [8675ffc](https://github.com/atomist/k8s-sdm/commit/8675ffc4d8f154ec08b0093ccce0c7bba732536d)
-   Print schedulers when k8s not found. [2723e33](https://github.com/atomist/k8s-sdm/commit/2723e33b732065761877babf3a0e6a34b4db9d73)
-   Execute goal if already scheduled. [257bd02](https://github.com/atomist/k8s-sdm/commit/257bd02404f4754dd0935a9ae5717d458c2432c6)
-   Persist removal of callback to goal event data. [8c2ab16](https://github.com/atomist/k8s-sdm/commit/8c2ab162443ddcf4790b5bfa17b89b620de96879)
-   Filter out sdm init containers. [c0a6b5e](https://github.com/atomist/k8s-sdm/commit/c0a6b5e631d0cd7bcb7c4f9c0c9feda9b7ddadd0)
-   Prefer container security context. [3cb1d2d](https://github.com/atomist/k8s-sdm/commit/3cb1d2d6c01ce5b2600df16e9df48463c8b73c1c)
-   Set defaultMode for k8s-sdm secret to 292. [af45b9d](https://github.com/atomist/k8s-sdm/commit/af45b9d4349fa493d020b2b0233f70ad9d5317a3)
-   Pure TypeScript, custom bucket, no logging. [05ee473](https://github.com/atomist/k8s-sdm/commit/05ee473a3958918050e891233974ad5fdee6a7bd)
-   Rearrange SDM configuration. [83691ec](https://github.com/atomist/k8s-sdm/commit/83691ecbb93c1b698348c6ebf9eaf15e4079ea7f)
-   Update sdm-pack-gcp. [97f3733](https://github.com/atomist/k8s-sdm/commit/97f37335f3d613dd331826ef7bae32321ae08ff7)
-   Give k8s-sdm more time to start up, remove fsgroup. [c20924f](https://github.com/atomist/k8s-sdm/commit/c20924f74b980d2f8d30ed95b75200d38bde20c8)
-   Improve local cache path and TypeScript config. [47680c7](https://github.com/atomist/k8s-sdm/commit/47680c77c742b9d3bc49e9495674715e12d3e0f2)

### Removed

-   Remove useless config default. [c628759](https://github.com/atomist/k8s-sdm/commit/c6287597cc7505ab9ec75de742a35a3bada58276)
-   Remove unused lodash import. [4073851](https://github.com/atomist/k8s-sdm/commit/4073851cee87307d2f2bb21e9bf1861bd9100b09)

### Fixed

-   Update sdm-pack-k8s for multi-spec yaml fix. [d375bab](https://github.com/atomist/k8s-sdm/commit/d375bab99583f59f2a49084633baab44576bfc57)
-   Update sdm-pack-k8s for scope fix. [005c8e6](https://github.com/atomist/k8s-sdm/commit/005c8e69685c7e08c2432b0cce2acc0ab441313f)
-   Update sdm-pack-k8s for namespace fix. [6b26526](https://github.com/atomist/k8s-sdm/commit/6b2652681aba245e76db7344de3d4d9e999ba528)
-   Ensure ATOMIST_GOAL_SCHEDULER set in k8s. [b424f00](https://github.com/atomist/k8s-sdm/commit/b424f0086d1d6839921bebb7a00c96bda2702255)
-   Fix multi-callback and goal state. [2902ff3](https://github.com/atomist/k8s-sdm/commit/2902ff32753ae1702e84519ebed5e9f4f96f4504)
-   Properly add extension packs. [067de0b](https://github.com/atomist/k8s-sdm/commit/067de0b7f43a95772ff254e981d3c8742dfcca5b)
-   Update sdm-pack-k8s for FulfillGoalOnRequested fix. [2e9d8a1](https://github.com/atomist/k8s-sdm/commit/2e9d8a1bf28ddc3c2e9ce0b95e9e1a723ee3eff3)

## [1.6.1](https://github.com/atomist/k8s-sdm/compare/1.6.0...1.6.1) - 2019-09-27

### Fixed

-   Update sdm-pack-k8s to remove error when sync repo not set and improve error & log messages. [8296445](https://github.com/atomist/k8s-sdm/commit/82964457944083d21b70e1c08567cdfb260b7728)

## [1.6.0](https://github.com/atomist/k8s-sdm/compare/1.5.0...1.6.0) - 2019-09-26

### Changed

-   Support multi-document YAML specs. [3f0085c](https://github.com/atomist/k8s-sdm/commit/3f0085cd32a7ab50de9fe077e1eb6bd59a054786)

### Fixed

-   Fix interval sync when commands not added. [f3e3ac9](https://github.com/atomist/k8s-sdm/commit/f3e3ac9676381a9cc3fca809baf5f4ad30a2f111)

## [1.5.0](https://github.com/atomist/k8s-sdm/compare/1.4.0...1.5.0) - 2019-08-15

### Changed

-   Run sync in worker. [84f01d3](https://github.com/atomist/k8s-sdm/commit/84f01d30af7bc40422ed981b68b5ebcdd27c8399)

test/empty.test.ts
/*
 * Copyright © 2020 Atomist, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as assert from "power-assert";

describe("empty", () => {
    it("is true", () => {
        assert(true);
    });
});


test/bucket.test.ts
/*
 * Copyright © 2020 Atomist, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as assert from "power-assert";
import { safeBucketName } from "../lib/bucket";

describe("bucket", () => {
	describe("safeBucketName", () => {
		it("should return the default bucket name", () => {
			[undefined, ""].forEach(n => {
				const b = safeBucketName(n);
				const e = "atomist-k8s-sdm";
				assert(b === e);
			});
		});

		it("should pad short names", () => {
			[
				{ b: "a", e: "a00" },
				{ b: "ab", e: "ab0" },
				{ b: "abc", e: "abc" },
				{ b: "abcd", e: "abcd" },
				{ b: "0", e: "000" },
				{ b: "97", e: "970" },
			].forEach(n => {
				const b = safeBucketName(n.b);
				assert(b === n.e);
			});
		});

		it("should truncate a long name", () => {
			const n =
				"some-very-long-bucket-name-that-is-too-long-to-be-an-actual-bucket-name-for-gcs";
			const b = safeBucketName(n);
			const e =
				"some-very-long-bucket-name-that-is-too-long-to-be-an-actual-buc";
			assert(b === e);
		});

		it("should safely truncate a long name", () => {
			const n =
				"some-very-long-bucket-name-that-is-too-long-to-be-an-actual-bu-cket-name-for-gcs";
			const b = safeBucketName(n);
			const e =
				"some-very-long-bucket-name-that-is-too-long-to-be-an-actual-bu";
			assert(b === e);
		});

		it("should return valid names", () => {
			[
				{ b: "@atomist/k8s-sdm", e: "atomist-k8s-sdm" },
				{ b: "@@@atomist//k8s--sdm@@", e: "atomist-k8s-sdm" },
				{ b: "ABC", e: "abc" },
				{ b: "192.168.0.1", e: "192-168-0-1" },
				{ b: "www.atomist.com", e: "www-atomist-com" },
			].forEach(n => {
				const b = safeBucketName(n.b);
				assert(b === n.e);
			});
		});

		it("should remove google", () => {
			[
				{ b: "googfoo", e: "atmfoo" },
				{ b: "goog", e: "atm" },
				{ b: "GOOG", e: "atm" },
				{ b: "some-google-bucket", e: "some-atomist-bucket" },
				{ b: "some-g00gle-bucket", e: "some-atomist-bucket" },
				{ b: "Some-g0oGlE-bucKet", e: "some-atomist-bucket" },
			].forEach(n => {
				const b = safeBucketName(n.b);
				assert(b === n.e);
			});
		});
	});
});


index.ts
/*
 * Copyright © 2020 Atomist, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { configure } from "@atomist/sdm/lib/core/machine/configure";
import { gcpSupport } from "@atomist/sdm/lib/pack/gcp/index";
import { k8sContainerFulfiller } from "@atomist/sdm/lib/pack/k8s/container";
import { kubernetesDeployFulfiller } from "@atomist/sdm/lib/pack/k8s/deploy/fulfiller";
import { k8sSupport } from "@atomist/sdm/lib/pack/k8s/k8s";
import { k8sGoalSchedulingSupport } from "@atomist/sdm/lib/pack/k8s/scheduler/goalScheduling";
import { machineOptions } from "./lib/configure";

/**
 * Provide a default configuration for this SDM.
 */
export const configuration = configure(async sdm => {
	sdm.addExtensionPacks(
		gcpSupport(),
		k8sGoalSchedulingSupport(),
		k8sSupport(),
	);
}, machineOptions);

/**
 * Fulfill containers goals sent to this SDM.
 */
export const containerFulfill = k8sContainerFulfiller();

/**
 * Fulfill Kubernetes deploy goals sent to this SDM.
 */
export const kubernetesDeployFulfill = kubernetesDeployFulfiller();


config.ts
/*
 * Copyright © 2020 Atomist, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Configuration } from "@atomist/automation-client";
import { configureHumio } from "@atomist/automation-client-ext-humio";
import { configureLogzio } from "@atomist/automation-client-ext-logzio";
import { k8sGoalSchedulingSupport } from "@atomist/sdm-core";
import { runningInK8s } from "@atomist/sdm-core/lib/goal/container/util";
import { gcpSupport } from "@atomist/sdm-pack-gcp";
import { k8sSupport } from "@atomist/sdm-pack-k8s";
import * as _ from "lodash";

export const K8sSupport = async (cfg: Configuration) => {
    if (runningInK8s()) {
        process.env.ATOMIST_GOAL_SCHEDULER = "kubernetes";
        const defaultCfg = {
            cluster: {
                workers: 1,
            },
            sdm: {
                k8s: {
                    job: {
                        cleanupInterval: 1000 * 60 * 10,
                    },
                },
                cache: {
                    bucket: "atm-demo-sdm-goal-cache-demo",
                    path: "demo-sdm-cache",
                },
                extensionPacks: [
                    gcpSupport(),
                    k8sGoalSchedulingSupport(),
                    k8sSupport({ registerCluster: true }),
                ],
            },
        };
        cfg.postProcessors = cfg.postProcessors || [];
        cfg.postProcessors.push(
            configureHumio,
            configureLogzio,
        );
        return _.defaultsDeep(cfg, defaultCfg);
    } else {
        const defaultCfg = {
            sdm: {
                cache: {
                    path: "/tmp/cache",
                },
            },
        };
        return _.defaultsDeep(cfg, defaultCfg);
    }
};


lib/bucket.ts
/*
 * Copyright © 2020 Atomist, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Make bucket name safe for Google Cloud Storage.
 * https://cloud.google.com/storage/docs/naming
 */
export function safeBucketName(bucket: string | undefined): string {
	if (!bucket) {
		return "atomist-k8s-sdm";
	}
	const safeBucket = bucket
		.toLowerCase()
		.replace(/[^-A-Za-z0-9]+/g, "-")
		.replace(/-+/g, "-")
		.replace(/^-/, "");
	const legalBucket = safeBucket
		.replace(/^goog/i, "atm")
		.replace(/g[o0][o0]gle/gi, "atomist");
	const lengthBucket = legalBucket.substring(0, 63).replace(/-$/, "");
	if (lengthBucket.length < 3) {
		return lengthBucket + "0".repeat(3 - lengthBucket.length);
	}
	return lengthBucket;
}


goals/kubernetes.ts
/*
 * Copyright © 2020 Atomist, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { GoalMaker } from "@atomist/sdm-core";
import { kubernetesDeployFulfiller } from "@atomist/sdm-pack-k8s";

export const kubernetesDeployFulfill: GoalMaker = async () => kubernetesDeployFulfiller();


lib/configure.ts
/*
 * Copyright © 2020 Atomist, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { SoftwareDeliveryMachineConfiguration } from "@atomist/sdm/lib/api/machine/SoftwareDeliveryMachineOptions";
import { runningInK8s } from "@atomist/sdm/lib/core/goal/container/util";
import { ConfigureMachineOptions } from "@atomist/sdm/lib/core/machine/configure";
import { LocalSoftwareDeliveryMachineConfiguration } from "@atomist/sdm/lib/core/machine/LocalSoftwareDeliveryMachineOptions";
import { defaultsDeep } from "lodash";
import { homedir } from "os";
import { join } from "path";
import { safeBucketName } from "./bucket";

/** Provide SDM configuration. */
async function configureSdm(
	cfg: LocalSoftwareDeliveryMachineConfiguration,
): Promise<LocalSoftwareDeliveryMachineConfiguration> {
	const defaultConfig: SoftwareDeliveryMachineConfiguration = {
		cluster: {
			workers: 1,
		},
		sdm: {
			goal: {
				timeout: 1000 * 60 * 20,
			},
			k8s: {
				job: {
					cleanupInterval: 1000 * 60 * 10,
				},
			},
			cache: {
				enabled: true,
			},
		},
	};
	let cacheConfig: SoftwareDeliveryMachineConfiguration;
	if (runningInK8s()) {
		process.env.ATOMIST_GOAL_SCHEDULER = "kubernetes";
		if (cfg.logging) {
			cfg.logging.color = false;
		}
		cacheConfig = {
			sdm: {
				cache: {
					bucket: safeBucketName(cfg.name),
					path: "k8s-sdm-cache",
				},
			},
		};
	} else {
		cacheConfig = {
			sdm: {
				cache: {
					path: join(homedir(), ".atomist", "cache"),
				},
			},
		};
	}
	return defaultsDeep(cfg, cacheConfig, defaultConfig);
}

/** SDM options. */
export const machineOptions: ConfigureMachineOptions = {
	preProcessors: [configureSdm],
};
